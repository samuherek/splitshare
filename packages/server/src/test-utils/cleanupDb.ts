import { getConnection } from 'typeorm';

export default async function cleanupDb() {
  await getConnection().query('DELETE FROM bill_user');
  await getConnection().query('DELETE FROM receipt');
  await getConnection().query('DELETE FROM bill_invite');
  await getConnection().query('DELETE FROM bill');
  await getConnection().query('DELETE FROM "user"');
}
