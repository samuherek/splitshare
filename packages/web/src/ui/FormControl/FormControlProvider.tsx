import React from 'react';

export type FormControlContextPropsKey =
  | 'disabled'
  | 'error'
  | 'required'
  | 'filled'
  | 'focused'
  | 'variant';

export type FormControlContextValue = {
  disabled: boolean;
  error: boolean;
  focused: boolean;
  required: boolean;
  variant: 'standard' | 'outline';
  filled: boolean;
  onFilled?: () => void;
  onEmpty?: () => void;
  onFocus?: () => void;
  onBlur?: () => void;
};

export interface FormControlProviderProps {
  value: FormControlContextValue;
}

const FormControlContext = React.createContext<
  FormControlContextValue | undefined
>(undefined);

function FormControlProvider(props: FormControlProviderProps) {
  const { value: valueProp, ...rest } = props;

  const value = React.useMemo(() => {
    return { ...valueProp };
  }, [valueProp]);

  return <FormControlContext.Provider value={value} {...rest} />;
}

function useFormControlCtx() {
  return React.useContext(FormControlContext);
}

export { FormControlContext, FormControlProvider, useFormControlCtx };
