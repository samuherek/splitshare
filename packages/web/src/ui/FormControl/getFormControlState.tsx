import {
  FormControlContextPropsKey,
  FormControlContextValue,
} from './FormControlProvider';

type Options = {
  props: object;
  states: FormControlContextPropsKey[];
  ctxFormControl?: FormControlContextValue;
};

export default function getFormControlState({
  props,
  states,
  ctxFormControl,
}: Options) {
  return states.reduce((acc, state) => {
    // @ts-ignore
    acc[state] = props[state];

    if (ctxFormControl) {
      // @ts-ignore
      if (typeof props[state] === 'undefined') {
        // @ts-ignore
        acc[state] = ctxFormControl[state];
      }
    }

    return acc;
  }, {});
}
