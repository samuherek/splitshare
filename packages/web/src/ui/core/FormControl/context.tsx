import React from 'react';

export interface CtxStateValues {
  disabled: boolean;
  error: boolean;
  focused: boolean;
  required: boolean;
  filled: boolean;
}

export interface CtxFunctions {
  onFilled?: () => void;
  onEmpty?: () => void;
  onFocus?: () => void;
  onBlur?: () => void;
}

export interface FormControlCtx extends CtxStateValues, CtxFunctions {}

export interface FormControlProviderProps {
  value: FormControlCtx;
  children: React.ReactNode;
}

const FormControlContext = React.createContext<FormControlCtx | undefined>(
  undefined
);

function useFormControlProvider(props: FormControlProviderProps) {
  const { value: valueProp } = props;

  const value = React.useMemo(() => ({ ...valueProp }), [valueProp]);

  return {
    props,
    value,
  };
}

function FormControlProvider(oriProps: FormControlProviderProps) {
  const { props, value } = useFormControlProvider(oriProps);
  return <FormControlContext.Provider value={value} {...props} />;
}

function useFormControlCtx() {
  return React.useContext(FormControlContext);
}

export { FormControlProvider, useFormControlCtx, useFormControlProvider };
