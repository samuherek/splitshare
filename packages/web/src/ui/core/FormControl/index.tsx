export { default } from './FormControl';
export * from './FormControl';
export * from './context';
