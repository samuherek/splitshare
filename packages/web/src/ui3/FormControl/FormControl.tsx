import clsx from 'clsx';
import React from 'react';
import styled from 'styled-components';
import { FormControlProvider } from './FormControlProvider';

interface Props {
  children: React.ReactNode;
  className?: string;
  as?: React.ReactNode;
  disabled?: boolean;
  error?: boolean;
  fullWidth?: boolean;
  required?: boolean;
  variant?: 'standard';
  // pretendFilled?: boolean;
}

export const classes = {
  root: 'SSFormControl',
  disabled: 'disabled',
  fullWidth: 'full-width',
  required: 'required',
  focused: 'focused',
  filled: 'filled',
  error: 'error',
};

const FormControlStyled = styled.div<{ fullWidth?: boolean }>`
  display: inline-flex;
  flex-direction: column;
  position: relative;
  min-width: 0;
  vertical-align: top;
  width: ${({ fullWidth }: any) => (fullWidth ? '100%' : 'auto')};
`;

const FormControl = React.forwardRef<Props, any>(function FormControl(
  props,
  ref
) {
  const {
    children,
    className,
    as = 'div',
    disabled = false,
    error = false,
    fullWidth = false,
    required = false,
    variant = 'standard',
    // pretendFilled = false,
    ...rest
  } = props;
  const [focused, setFocused] = React.useState(false);
  const [filled, setFilled] = React.useState(false);

  const handleDirty = () => {
    if (!filled) {
      setFilled(true);
    }
  };

  const handleClean = () => {
    if (filled) {
      setFilled(false);
    }
  };

  const handleFocus = () => {
    setFocused(true);
  };

  const handleBlur = () => {
    setFocused(false);
  };

  const childContext = {
    disabled,
    error,
    required,
    variant,
    focused,
    filled,
    onFilled: handleDirty,
    onEmpty: handleClean,
    onFocus: handleFocus,
    onBlur: handleBlur,
  };

  return (
    // @ts-ignore
    <FormControlProvider value={childContext}>
      <FormControlStyled
        as={as}
        className={clsx(className, classes.root, {
          [classes.disabled]: disabled,
          [classes.fullWidth]: fullWidth,
          [classes.required]: required,
          [classes.focused]: focused,
          [classes.filled]: filled,
          [classes.error]: error,
        })}
        ref={ref}
        fullWidth={fullWidth}
        // pretendFilled={pretendFilled}
        {...rest}
      >
        {children}
      </FormControlStyled>
    </FormControlProvider>
  );
});

export default FormControl;
