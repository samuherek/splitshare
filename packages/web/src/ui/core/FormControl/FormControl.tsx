import React from 'react';
import cx from 'clsx';
import { FormControlProvider } from './context';
import { UICoreBaseProps } from '../types';
import isUiElement from '../../utils/isUiElement';
import { isFilled } from '../InputCore/utils';

export interface FormControlProps extends UICoreBaseProps {
  disabled: boolean;
  error: boolean;
  focused: boolean;
  // hiddenLabel: boolean,
  required: boolean;
}

export const classes = {
  root: 'FormControl',
};

export function useFormControl(props: FormControlProps, ref: React.Ref<any>) {
  const {
    className,
    component,
    disabled,
    error,
    focused: visuallyFocused,
    required,
    children,
  } = props;

  const [_focused, setFocused] = React.useState(false);

  const [filled, setFilled] = React.useState(() => {
    // We need to iterate through the children and find the Input in order
    // to fully support server-side rendering.
    let initialFilled = false;

    if (children) {
      React.Children.forEach(children, child => {
        if (!isUiElement(child, ['Input', 'Select'])) {
          return;
        }

        // @ts-ignore
        if (isFilled(child.props, true)) {
          initialFilled = true;
        }
      });
    }

    return initialFilled;
  });

  const focused = visuallyFocused !== undefined ? visuallyFocused : _focused;

  if (disabled && focused) {
    setFocused(false);
  }

  const onFilled = React.useCallback(() => {
    setFilled(true);
  }, []);

  const onEmpty = React.useCallback(() => {
    setFilled(false);
  }, []);

  const onBlur = React.useCallback(() => {
    setFocused(false);
  }, []);

  const onFocus = React.useCallback(() => {
    setFocused(true);
  }, []);

  const context = {
    disabled,
    focused,
    error,
    filled,
    required,
    onEmpty,
    onFilled,
    onBlur,
    onFocus,
  };

  return {
    props: {
      ...props,
      className: cx(className, classes.root),
    },
    ref,
    Component: component || 'div',
    context,
  };
}

const FormControl = React.forwardRef<any, FormControlProps>((p, r) => {
  const { props, ref, context, Component } = useFormControl(p, r);
  // FIXME: figure out how to type this
  const C: any = Component;

  return (
    <FormControlProvider value={context}>
      <C ref={ref} {...props} />
    </FormControlProvider>
  );
});

export default FormControl;
