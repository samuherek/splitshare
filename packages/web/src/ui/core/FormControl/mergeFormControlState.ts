import { CtxStateValues, FormControlCtx } from './context';

type Options = {
  props: Partial<CtxStateValues>;
  states: Array<keyof CtxStateValues>;
  uiFormControlCtx?: FormControlCtx;
};

export default function mergeFormControlState({
  props,
  states,
  uiFormControlCtx,
}: Options) {
  return states.reduce<Partial<CtxStateValues>>((acc, state) => {
    acc[state] = props[state];

    if (uiFormControlCtx) {
      if (typeof props[state] === 'undefined') {
        acc[state] = uiFormControlCtx[state];
      }
    }

    return acc;
  }, {});
}
