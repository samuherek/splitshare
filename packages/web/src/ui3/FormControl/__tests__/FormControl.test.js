import React from 'react';
import { render } from '../../../test-utils/test-utils';
import FormControl, { classes } from '../FormControl';

it('should match snapshots', () => {
  const { container } = render(<FormControl />);
  expect(container.firstChild).toMatchSnapshot();
});

it('should render <FormControl /> correctly', () => {
  const { container } = render(<FormControl />);
  const fc = container.firstChild;

  expect(fc).toHaveClass(classes.root);
  expect(fc).not.toHaveClass(classes.disabled);
  expect(fc).not.toHaveClass(classes.error);
  expect(fc).not.toHaveClass(classes.filled);
  expect(fc).not.toHaveClass(classes.focused);
  expect(fc).not.toHaveClass(classes.required);
  expect(fc).not.toHaveClass(classes.fullWidth);
  expect(fc).toHaveStyleRule('position', 'relative');
  expect(fc).toHaveStyleRule('display', 'inline-flex');
  expect(fc).toHaveStyleRule('flex-direction', 'column');
});

test('prop: disabled should have class', () => {
  const { container } = render(<FormControl disabled />);
  const fc = container.firstChild;
  expect(fc).toHaveClass(classes.disabled);
});

test('prop: error should have class', () => {
  const { container } = render(<FormControl error />);
  const fc = container.firstChild;
  expect(fc).toHaveClass(classes.error);
});

test('prop: required should have class', () => {
  const { container } = render(<FormControl required />);
  const fc = container.firstChild;
  expect(fc).toHaveClass(classes.required);
});

test('prop: fullWidth should have class', () => {
  const { container } = render(<FormControl fullWidth />);
  const fc = container.firstChild;
  expect(fc).toHaveClass(classes.fullWidth);
});
