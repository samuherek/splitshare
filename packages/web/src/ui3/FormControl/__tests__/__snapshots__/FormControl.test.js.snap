// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`should match snapshots 1`] = `
.c0 {
  display: -webkit-inline-box;
  display: -webkit-inline-flex;
  display: -ms-inline-flexbox;
  display: inline-flex;
  -webkit-flex-direction: column;
  -ms-flex-direction: column;
  flex-direction: column;
  position: relative;
  min-width: 0;
  vertical-align: top;
  width: auto;
}

<div
  class="SSFormControl c0"
/>
`;
